File: src/tts/utils/pcm-to-opus.ts
/**
 * PCM → Opus 流式编码器
 *
 * 将 PCM 音频流流式编码为 Opus 数据包流（可选 OGG 容器封装）。
 * 依赖 prism-media（作为 optionalDependency），使用 libopus 原生编码能力。
 *
 * 核心机制：
 * 1. 从 TTS 流式拉取 PCM chunk（AsyncIterable<TTSStreamChunk> 或裸 Uint8Array 流）
 * 2. 帧缓冲区累积 PCM 数据，凑够一整帧后送入 Opus 编码器
 * 3. 编码器输出 Opus packet 后立即 yield 给消费者（真正的流式，不等待全部数据）
 * 4. 流结束时尾部不足一帧的数据用零填充
 *
 * @module pcm-to-opus
 */

import { Buffer } from 'node:buffer';
import { createOggMuxerWithEos } from '@/asr/utils/ogg-muxer';
import type { TTSStreamChunk } from '@/types/tts';

/** Opus 支持的帧时长（毫秒） */
const VALID_FRAME_DURATIONS_MS = [2.5, 5, 10, 20, 40, 60];

/**
 * PCM → Opus 流式编码选项
 */
export interface PcmToOpusOptions {
  /** PCM 采样率（Hz），默认 24000（与 Doubao TTS PCM 输出一致） */
  sampleRate?: number;
  /** 声道数，默认 1（单声道） */
  channels?: number;
  /**
   * Opus 帧时长（毫秒），默认 60
   *
   * 可选值: 2.5, 5, 10, 20, 40, 60
   * - 20ms: 标准值，延迟与压缩率平衡
   * - 60ms: 高压缩率，适合 TTS 离线场景
   */
  frameDurationMs?: number;
  /** PCM 位深（bytes per sample），默认 2（16-bit） */
  bytesPerSample?: number;
  /**
   * 是否封装为 OGG 容器格式，默认 false
   *
   * 设为 true 时输出可直接播放的 OGG 流，
   * 内部复用 createOggMuxerWithEos 实现。
   */
  ogg?: boolean | { encoder?: string };
}

/**
 * 将 PCM 音频流流式编码为 Opus 数据包流
 *
 * @param pcmStream PCM 音频数据流
 *   - AsyncIterable<TTSStreamChunk>: TTS speak() 的标准返回类型，自动提取 audioChunk
 *   - AsyncIterable<Uint8Array> | AsyncIterable<Buffer>: 原始 PCM 字节流
 * @param options 编码选项
 * @returns Opus 数据包流（AsyncIterable<Buffer>）
 *   - 裸 Opus 模式（默认）: 每个 Buffer 是一个独立的 Opus 编码帧
 *   - OGG 模式（{ ogg: true }）: 每个 Buffer 是一个 OGG 页面，可直接写入文件播放
 *
 * @example
 * ```typescript
 * import { createTTS, pcmToOpus } from 'univoice';
 *
 * const tts = createTTS({ provider: 'doubao', format: 'pcm', sampleRate: 24000 });
 * const pcmStream = tts.speak('你好世界', { stream: true });
 *
 * // 裸 Opus 包流
 * for await (const packet of pcmToOpus(pcmStream)) {
 *   ws.send(packet);
 * }
 *
 * // OGG 封装流（可直接播放）
 * for await (const page of pcmToOpus(pcmStream, { ogg: true })) {
 *   // 写入文件或发送到 WebSocket
 * }
 * ```
 */
export async function* pcmToOpus(
  pcmStream: AsyncIterable<TTSStreamChunk> | AsyncIterable<Uint8Array> | AsyncIterable<Buffer>,
  options?: PcmToOpusOptions
): AsyncIterable<Buffer> {
  // ====== 参数解析与校验 ======
  const {
    sampleRate = 24000,
    channels = 1,
    frameDurationMs = 60,
    bytesPerSample = 2,
    ogg: oggEnabled,
  } = options || {};

  if (sampleRate <= 0) {
    throw new RangeError('sampleRate must be a positive number');
  }
  if (channels <= 0) {
    throw new RangeError('channels must be a positive number');
  }
  if (!VALID_FRAME_DURATIONS_MS.includes(frameDurationMs)) {
    throw new RangeError(`frameDurationMs must be one of: ${VALID_FRAME_DURATIONS_MS.join(', ')}`);
  }
  if (bytesPerSample !== 1 && bytesPerSample !== 2) {
    throw new RangeError('bytesPerSample must be 1 or 2');
  }

  const frameSizeSamples = Math.round((sampleRate / 1000) * frameDurationMs);
  const frameSizeBytes = frameSizeSamples * bytesPerSample;

  // ====== 动态导入 prism-media ======
  let prismMedia: typeof import('prism-media');
  try {
    prismMedia = await import('prism-media');
  } catch {
    throw new Error(
      'prism-media is required for Opus encoding but is not installed. ' +
        'Install it with: pnpm add prism-media'
    );
  }

  // ====== 创建 Opus 编码器（Transform Stream）======
  const encoder = new prismMedia.opus.Encoder({
    frameSize: frameSizeSamples,
    channels,
    rate: sampleRate,
  });

  // ====== 输出队列与状态 ======
  const outputQueue: Buffer[] = [];
  let encodeError: Error | null = null;
  let outputDone = false;

  // 条件变量：用于在输出队列为空且未完成时阻塞主循环，
  // 当有新数据入队或流结束时唤醒。
  // 无论裸 Opus 模式还是 OGG 模式，都通过此机制统一唤醒，避免事件丢失。
  let wakeUpResolver: (() => void) | null = null;

  function wakeUp(): void {
    if (wakeUpResolver) {
      const r = wakeUpResolver;
      wakeUpResolver = null;
      r();
    }
  }

  // ====== OGG 模式初始化 ======
  const useOgg = !!oggEnabled;
  const oggEncoderName = typeof oggEnabled === 'object' ? oggEnabled.encoder : undefined;

  let oggMuxer: ReturnType<typeof createOggMuxerWithEos> | null = null;
  let pushIterable: PushIterable<Buffer> | null = null;

  if (useOgg) {
    // 创建可手动推送的 AsyncIterable 适配器，将 encoder 的 data 事件桥接为 AsyncIterable
    pushIterable = new PushIterable<Buffer>();
    oggMuxer = createOggMuxerWithEos(pushIterable.asyncIterable, {
      sampleRate,
      channels,
      frameSizeMs: frameDurationMs,
      encoder: oggEncoderName ?? 'univoice',
    });
    // 从 OGG muxer 的流中拉取页面并送入输出队列
    (async () => {
      try {
        for await (const page of oggMuxer.stream) {
          outputQueue.push(page);
          wakeUp();
        }
      } catch (err) {
        if (!encodeError) {
          encodeError = err instanceof Error ? err : new Error(String(err));
        }
      } finally {
        outputDone = true;
        wakeUp();
      }
    })();
  }

  // 收集编码输出（裸 Opus 模式直接入队；OGG 模式通过 PushIterable 送入 muxer）
  encoder.on('data', (chunk: Buffer) => {
    if (useOgg && pushIterable) {
      pushIterable.push(chunk);
    } else {
      outputQueue.push(chunk);
      wakeUp();
    }
  });

  encoder.on('error', (err: Error) => {
    encodeError = err;
    if (!useOgg) {
      outputDone = true;
    }
    wakeUp();
  });

  // 裸 Opus 模式的结束信号
  if (!useOgg) {
    encoder.on('end', () => {
      outputDone = true;
      wakeUp();
    });
  }

  // ====== 后台任务：从 PCM 流拉取数据并送入编码器 ======
  const writePromise = (async () => {
    let pcmBuffer = Buffer.alloc(0);

    try {
      for await (const chunk of pcmStream) {
        if (encodeError) break;

        // 类型收窄：TTSStreamChunk 提取 audioChunk，其他直接使用
        const rawData = extractPcmData(chunk);
        const data = Buffer.isBuffer(rawData) ? rawData : Buffer.from(rawData);

        // 追加到帧缓冲区
        pcmBuffer = Buffer.concat([pcmBuffer, data]);

        // 凑够一帧就编码
        while (pcmBuffer.length >= frameSizeBytes) {
          const frame = pcmBuffer.subarray(0, frameSizeBytes);
          pcmBuffer = pcmBuffer.subarray(frameSizeBytes);
          encoder.write(frame);
        }
      }

      // 尾部不足一帧的数据用零填充（静音填充）
      if (pcmBuffer.length > 0) {
        const paddedFrame = Buffer.alloc(frameSizeBytes);
        pcmBuffer.copy(paddedFrame);
        encoder.write(paddedFrame);
      }
    } catch (err) {
      encodeError = err instanceof Error ? err : new Error(String(err));
    } finally {
      // 结束编码器（刷新内部缓冲区）
      encoder.end();

      // OGG 模式：通知 muxer 发送 EOS 页面
      if (useOgg && oggMuxer) {
        oggMuxer.finish();
        if (pushIterable) {
          pushIterable.finish();
        }
      }
    }
  })();

  // 等待后台任务启动
  await new Promise((resolve) => setImmediate(resolve));

  // ====== 主循环：从队列 yield 数据给消费者 ======
  while (!outputDone || outputQueue.length > 0) {
    if (encodeError) {
      // 等待后台任务完成以确保资源清理
      await writePromise.catch(() => {});
      throw encodeError;
    }

    if (outputQueue.length > 0) {
      const item = outputQueue.shift();
      if (item) yield item;
    } else if (!outputDone) {
      // 等待有新数据入队或流结束（通过 wakeUp 条件变量唤醒）
      await new Promise<void>((resolve) => {
        wakeUpResolver = resolve;
      });
    }
  }

  // 等待后台写入任务完成
  await writePromise;

  if (encodeError) {
    throw encodeError;
  }
}

// ==================== 内部工具 ====================

/**
 * 从流式 chunk 中提取原始 PCM 数据
 *
 * @internal
 */
function extractPcmData(chunk: TTSStreamChunk | Uint8Array | Buffer): Uint8Array | Buffer {
  if ('audioChunk' in chunk && chunk.audioChunk instanceof Uint8Array) {
    return chunk.audioChunk;
  }
  return chunk as Uint8Array | Buffer;
}

/**
 * 可手动推送数据的 AsyncIterable 适配器
 *
 * 用于将事件驱动（如 encoder.on('data')）的数据源转换为 AsyncIterable<Buffer>，
 * 以便接入 createOggMuxerWithEos 等 AsyncIterable 消费者。
 *
 * @internal
 */
class PushIterable<T> implements AsyncIterable<T> {
  private queue: T[] = [];
  private done = false;
  private error: Error | null = null;
  private pendingResolver: ((result: IteratorResult<T>) => void) | null = null;

  push(value: T): void {
    this.queue.push(value);
    if (this.pendingResolver) {
      const resolver = this.pendingResolver;
      this.pendingResolver = null;
      const shifted = this.queue.shift();
      resolver({ value: shifted as T, done: false });
    }
  }

  finish(err?: Error): void {
    if (err) {
      this.error = err;
    }
    this.done = true;
    if (this.pendingResolver) {
      const resolver = this.pendingResolver;
      this.pendingResolver = null;
      if (this.error) {
        resolver(Promise.reject(this.error) as unknown as IteratorResult<T>);
      } else {
        resolver({ value: undefined as T, done: true });
      }
    }
  }

  get asyncIterable(): AsyncIterable<T> {
    return this;
  }

  [Symbol.asyncIterator](): AsyncIterator<T> {
    return {
      next: (): Promise<IteratorResult<T>> => {
        if (this.queue.length > 0) {
          const shifted = this.queue.shift();
          return Promise.resolve({
            value: shifted as T,
            done: false,
          });
        }
        if (this.error) {
          const err = this.error;
          this.error = null;
          return Promise.reject(err);
        }
        if (this.done) {
          return Promise.resolve({
            value: undefined as T,
            done: true,
          });
        }
        return new Promise((resolve) => {
          this.pendingResolver = resolve;
        });
      },
    };
  }
}

File: src/tts/utils/pcm_to_opus.rs
/// PCM → Opus 流式编码器
///
/// 将 PCM 音频流流式编码为 Opus 数据包流（可选 OGG 容器封装）。
/// 依赖 `opus2` crate，编译时自动构建 libopus。
///
/// 对应 TypeScript 的 `pcmToOpus` 函数。
///
/// # Feature Gate
///
/// 本模块仅在 `opus-encoder` feature 启用时可用：
/// ```toml
/// univoice = { features = ["opus-encoder"] }
/// ```
use std::pin::Pin;

use futures_util::Stream;
use futures_util::StreamExt;
use opus2::{self, Application, Channels};

use crate::asr::utils::OggMuxer;
use crate::asr::utils::OggMuxerOptions;
use crate::tts::error::TtsError;
use crate::tts::types::TtsAudioStream;
#[cfg(test)]
use crate::tts::types::TtsStreamChunk;

/// PCM → Opus 编码后的流类型
type OpusStream = Pin<Box<dyn Stream<Item = Result<Vec<u8>, TtsError>> + Send>>;

/// Opus 支持的帧时长（毫秒）
const VALID_FRAME_DURATIONS_MS: &[u32] = &[2, 5, 10, 20, 40, 60, 120];

/// PCM → Opus 流式编码选项
#[derive(Debug, Clone)]
pub struct PcmToOpusOptions {
    /// PCM 采样率（Hz，默认 24000）
    pub sample_rate: u32,
    /// 声道数（默认 1）
    pub channels: u8,
    /// Opus 帧时长（毫秒，默认 60）
    ///
    /// 可选值: 2, 5, 10, 20, 40, 60, 120
    /// - 20ms: 标准值，延迟与压缩率平衡
    /// - 60ms: 高压缩率，适合 TTS 离线场景
    pub frame_duration_ms: u32,
    /// PCM 采样位深（bytes per sample，默认 2，即 16-bit）
    pub bytes_per_sample: usize,
    /// 是否封装为 OGG 容器格式（默认 false）
    pub ogg: bool,
    /// OGG 编码器名称（ogg=true 时有效）
    pub ogg_encoder: Option<String>,
    /// Opus 编码复杂度（0-10，默认 10）
    pub complexity: Option<i32>,
}

impl Default for PcmToOpusOptions {
    fn default() -> Self {
        Self {
            sample_rate: 24000,
            channels: 1,
            frame_duration_ms: 60,
            bytes_per_sample: 2,
            ogg: false,
            ogg_encoder: None,
            complexity: Some(10),
        }
    }
}

/// 将 PCM 音频流流式编码为 Opus 数据包流
///
/// 对应 TypeScript 的 `pcmToOpus` 函数。
///
/// # 参数
/// - `pcm_stream`: TTS 流式音频输出（包含 PCM 数据块）
/// - `options`: 编码选项
///
/// # 返回值
/// - Opus 数据包流（每个元素是一个 Opus 编码帧或 OGG 页面）
///
/// # 错误
/// - 如果参数不合法
/// - 如果 Opus 编码失败
///
/// # 示例
/// ```rust,ignore
/// use univoice::tts::utils::pcm_to_opus;
///
/// let opus_stream = pcm_to_opus(tts_stream, Default::default()).await?;
/// ```
pub fn pcm_to_opus(
    pcm_stream: TtsAudioStream,
    options: PcmToOpusOptions,
) -> Result<OpusStream, TtsError> {
    // ====== 参数校验 ======
    validate_options(&options)?;

    let frame_size_samples =
        ((options.sample_rate as u64 * options.frame_duration_ms as u64) / 1000) as usize;
    let frame_size_bytes = frame_size_samples * options.bytes_per_sample;

    // ====== 创建 Opus 编码器 ======
    let channels_enum = match options.channels {
        1 => Channels::Mono,
        2 => Channels::Stereo,
        n => {
            return Err(TtsError::InvalidParameter(format!(
                "Unsupported channels: {n} (only 1 or 2)"
            )));
        }
    };

    let mut encoder =
        opus2::Encoder::new(options.sample_rate, channels_enum, Application::Audio)
            .map_err(|e| TtsError::Other(format!("Failed to create Opus encoder: {e}")))?;

    // 设置编码复杂度
    if let Some(complexity) = options.complexity {
        if let Err(e) = encoder.set_complexity(complexity) {
            return Err(TtsError::Other(format!(
                "Failed to set Opus complexity: {e}"
            )));
        }
    }

    // ====== 准备 OGG Muxer（如果启用）======
    let use_ogg = options.ogg;
    let _ogg_encoder_name = options
        .ogg_encoder
        .clone()
        .unwrap_or_else(|| "univoice-rs".to_string());

    Ok(Box::pin(async_stream::try_stream! {
        let mut pcm_buffer: Vec<u8> = Vec::new();
        let mut ogg_muxer: Option<OggMuxer> = None;

        if use_ogg {
            ogg_muxer = Some(OggMuxer::new(OggMuxerOptions {
                sample_rate: options.sample_rate,
                channels: options.channels,
                frame_size_ms: options.frame_duration_ms,
            }));
        }

        tokio::pin!(pcm_stream);

        while let Some(chunk_result) = pcm_stream.next().await {
            let chunk = chunk_result?;
            // 累积 PCM 数据
            pcm_buffer.extend_from_slice(&chunk.audio_chunk);

            // 当缓冲区有足够数据构成一帧时，循环编码
            while pcm_buffer.len() >= frame_size_bytes {
                let frame = pcm_buffer[..frame_size_bytes].to_vec();
                pcm_buffer = pcm_buffer[frame_size_bytes..].to_vec();

                // PCM i16le 转 Opus 期望的 &[i16]
                let pcm_i16: Vec<i16> = frame
                    .chunks_exact(2)
                    .map(|b| i16::from_le_bytes([b[0], b[1]]))
                    .collect();

                let mut opus_packet = vec![0u8; 4000]; // Opus 最大包大小
                let encoded_len = encoder
                    .encode(&pcm_i16, &mut opus_packet)
                    .map_err(|e| TtsError::Other(format!("Opus encode error: {e}")))?;

                opus_packet.truncate(encoded_len);

                // OGG 封装或裸 Opus
                if let Some(ref mut muxer) = ogg_muxer {
                    let pages = muxer.push_packet(&opus_packet);
                    for page in pages {
                        yield page;
                    }
                } else {
                    yield opus_packet;
                }
            }
        }

        // 尾部不足一帧的数据用零填充
        if !pcm_buffer.is_empty() {
            let mut padded = vec![0u8; frame_size_bytes];
            let copy_len = pcm_buffer.len().min(frame_size_bytes);
            padded[..copy_len].copy_from_slice(&pcm_buffer[..copy_len]);

            let pcm_i16: Vec<i16> = padded
                .chunks_exact(2)
                .map(|b| i16::from_le_bytes([b[0], b[1]]))
                .collect();

            let mut opus_packet = vec![0u8; 4000];
            let encoded_len = encoder
                .encode(&pcm_i16, &mut opus_packet)
                .map_err(|e| TtsError::Other(format!("Opus encode error: {e}")))?;

            opus_packet.truncate(encoded_len);

            if let Some(ref mut muxer) = ogg_muxer {
                let pages = muxer.push_packet(&opus_packet);
                for page in pages {
                    yield page;
                }
            } else {
                yield opus_packet;
            }
        }

        // OGG 结束
        if let Some(ref mut muxer) = ogg_muxer {
            if let Some(eos_page) = muxer.finish() {
                yield eos_page;
            }
        }
    }))
}

/// 验证 PCM→Opus 编码选项
fn validate_options(options: &PcmToOpusOptions) -> Result<(), TtsError> {
    if options.sample_rate == 0 {
        return Err(TtsError::InvalidParameter(
            "sample_rate must be a positive number".into(),
        ));
    }
    if options.channels != 1 && options.channels != 2 {
        return Err(TtsError::InvalidParameter(
            "channels must be 1 (mono) or 2 (stereo)".into(),
        ));
    }
    if !VALID_FRAME_DURATIONS_MS.contains(&options.frame_duration_ms) {
        return Err(TtsError::InvalidParameter(format!(
            "frame_duration_ms must be one of: {:?}",
            VALID_FRAME_DURATIONS_MS
        )));
    }
    if options.bytes_per_sample != 1 && options.bytes_per_sample != 2 {
        return Err(TtsError::InvalidParameter(
            "bytes_per_sample must be 1 or 2".into(),
        ));
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use futures_util::stream;

    #[test]
    fn test_validate_options_default() {
        let opts = PcmToOpusOptions::default();
        assert!(validate_options(&opts).is_ok());
    }

    #[test]
    fn test_validate_options_invalid_sample_rate() {
        let opts = PcmToOpusOptions {
            sample_rate: 0,
            ..Default::default()
        };
        assert!(validate_options(&opts).is_err());
    }

    #[test]
    fn test_validate_options_invalid_channels() {
        let opts = PcmToOpusOptions {
            channels: 3,
            ..Default::default()
        };
        assert!(validate_options(&opts).is_err());
    }

    #[test]
    fn test_validate_options_invalid_frame_duration() {
        let opts = PcmToOpusOptions {
            frame_duration_ms: 7,
            ..Default::default()
        };
        assert!(validate_options(&opts).is_err());
    }

    #[test]
    fn test_validate_options_invalid_bytes_per_sample() {
        let opts = PcmToOpusOptions {
            bytes_per_sample: 4,
            ..Default::default()
        };
        assert!(validate_options(&opts).is_err());
    }

    #[tokio::test]
    async fn test_pcm_to_opus_empty_stream() {
        let stream: TtsAudioStream = Box::pin(stream::iter(vec![]));
        let opts = PcmToOpusOptions::default();
        let mut opus_stream = pcm_to_opus(stream, opts).unwrap();

        use futures_util::StreamExt;
        let results: Vec<Result<Vec<u8>, TtsError>> = opus_stream.by_ref().collect().await;
        // 空流应该没有输出（无 PCM 数据送入编码器）
        assert!(results.is_empty());
    }

    #[tokio::test]
    async fn test_pcm_to_opus_with_audio_data() {
        // 产生 60ms @ 24kHz 16bit mono 的 PCM 数据
        // frame_size_samples = 24000 * 60 / 1000 = 1440 samples
        // frame_size_bytes = 1440 * 2 = 2880 bytes
        let frame_bytes = 2880;
        let pcm_data = vec![0u8; frame_bytes]; // 静音（全零）

        let chunks: Vec<Result<TtsStreamChunk, TtsError>> = vec![Ok(TtsStreamChunk {
            audio_chunk: pcm_data,
        })];

        let stream: TtsAudioStream = Box::pin(stream::iter(chunks));
        let opts = PcmToOpusOptions::default();
        let mut opus_stream = pcm_to_opus(stream, opts).unwrap();

        use futures_util::StreamExt;
        let results: Vec<Result<Vec<u8>, TtsError>> = opus_stream.by_ref().collect().await;

        // 应该有编码结果
        assert!(!results.is_empty(), "Should produce Opus packets");
        if let Ok(packet) = &results[0] {
            assert!(!packet.is_empty(), "Opus packet should not be empty");
            // 静音编码后的 Opus 包应该很小（约 10-20 bytes）
            assert!(
                packet.len() < 100,
                "Silence Opus packet should be small: {}",
                packet.len()
            );
        }
    }

    #[tokio::test]
    async fn test_pcm_to_opus_ogg_mode() {
        let frame_bytes = 2880; // 60ms @ 24kHz 16bit mono
        let pcm_data = vec![0u8; frame_bytes];

        let chunks: Vec<Result<TtsStreamChunk, TtsError>> = vec![Ok(TtsStreamChunk {
            audio_chunk: pcm_data,
        })];

        let stream: TtsAudioStream = Box::pin(stream::iter(chunks));
        let opts = PcmToOpusOptions {
            ogg: true,
            ..Default::default()
        };

        let mut opus_stream = pcm_to_opus(stream, opts).unwrap();

        use futures_util::StreamExt;
        let results: Vec<Result<Vec<u8>, TtsError>> = opus_stream.by_ref().collect().await;

        assert!(!results.is_empty(), "Should produce OGG pages");

        // OGG 模式下的第一页应该是 OpusHead（BOS）
        if let Ok(page) = &results[0] {
            assert_eq!(&page[0..4], b"OggS", "Should be OGG page");
            assert_eq!(page[5], 2, "First page should have BOS flag");
        }
    }

    #[tokio::test]
    async fn test_pcm_to_opus_multi_frame() {
        // 3 整帧 PCM 数据
        let frame_bytes = 2880;
        let pcm_data = vec![0u8; frame_bytes * 3];

        let chunks: Vec<Result<TtsStreamChunk, TtsError>> = vec![Ok(TtsStreamChunk {
            audio_chunk: pcm_data,
        })];

        let stream: TtsAudioStream = Box::pin(stream::iter(chunks));
        let mut opus_stream = pcm_to_opus(stream, PcmToOpusOptions::default()).unwrap();

        use futures_util::StreamExt;
        let results: Vec<Result<Vec<u8>, TtsError>> = opus_stream.by_ref().collect().await;

        // 3 帧应产生 3 个 Opus 包
        assert_eq!(results.len(), 3, "3 frames should produce 3 packets");
        for result in &results {
            assert!(result.is_ok());
            assert!(!result.as_ref().unwrap().is_empty());
        }
    }

    #[tokio::test]
    async fn test_pcm_to_opus_partial_last_frame() {
        // 2.5 帧 PCM 数据 - 尾部半帧应被零填充
        let frame_bytes = 2880;
        let pcm_data = vec![0xFFu8; frame_bytes * 2 + frame_bytes / 2];

        let chunks: Vec<Result<TtsStreamChunk, TtsError>> = vec![Ok(TtsStreamChunk {
            audio_chunk: pcm_data,
        })];

        let stream: TtsAudioStream = Box::pin(stream::iter(chunks));
        let mut opus_stream = pcm_to_opus(stream, PcmToOpusOptions::default()).unwrap();

        use futures_util::StreamExt;
        let results: Vec<Result<Vec<u8>, TtsError>> = opus_stream.by_ref().collect().await;

        // 2 整帧 + 1 填充帧 = 3 个包
        assert_eq!(results.len(), 3, "2.5 frames should produce 3 packets");
    }

    #[tokio::test]
    async fn test_pcm_to_opus_different_sample_rate() {
        // 20ms @ 16kHz 16bit mono = 16000 * 20 / 1000 * 2 = 640 bytes
        let frame_bytes = 640;
        let pcm_data = vec![0u8; frame_bytes];

        let chunks: Vec<Result<TtsStreamChunk, TtsError>> = vec![Ok(TtsStreamChunk {
            audio_chunk: pcm_data,
        })];

        let stream: TtsAudioStream = Box::pin(stream::iter(chunks));
        let opts = PcmToOpusOptions {
            sample_rate: 16000,
            frame_duration_ms: 20,
            ..Default::default()
        };

        let mut opus_stream = pcm_to_opus(stream, opts).unwrap();

        use futures_util::StreamExt;
        let results: Vec<Result<Vec<u8>, TtsError>> = opus_stream.by_ref().collect().await;

        assert!(!results.is_empty(), "16kHz 20ms should produce packets");
        for result in &results {
            assert!(result.is_ok());
        }
    }

    #[tokio::test]
    async fn test_pcm_to_opus_chunks_across_frames() {
        // 每个 chunk 小于一帧，需要累积
        let frame_bytes = 2880;
        let half_frame = frame_bytes / 2;

        let chunks: Vec<Result<TtsStreamChunk, TtsError>> = vec![
            Ok(TtsStreamChunk {
                audio_chunk: vec![0u8; half_frame],
            }),
            Ok(TtsStreamChunk {
                audio_chunk: vec![0u8; half_frame],
            }),
            Ok(TtsStreamChunk {
                audio_chunk: vec![0u8; half_frame],
            }),
            Ok(TtsStreamChunk {
                audio_chunk: vec![0u8; half_frame],
            }),
        ];

        let stream: TtsAudioStream = Box::pin(stream::iter(chunks));
        let mut opus_stream = pcm_to_opus(stream, PcmToOpusOptions::default()).unwrap();

        use futures_util::StreamExt;
        let results: Vec<Result<Vec<u8>, TtsError>> = opus_stream.by_ref().collect().await;

        // 4 x 1440 = 5760 bytes = 2 整帧，应产生 2 个包
        assert_eq!(
            results.len(),
            2,
            "Cross-chunk accumulation should produce 2 packets"
        );
    }
}
